test/integration/volume/attach_detach_test.go

```diff
@@ -138,6 +142,107 @@ var defaultTimerConfig = attachdetach.TimerConfig{
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins=ServiceAccount"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+	namespaceName := "test-volume-terminating-detach"
+	nodeName := "node-sandbox"
+	testClient, ctrl, pvCtrl, gcCtrl, informers := createAdClients(t, server, defaultSyncPeriod, defaultTimerConfig)
+	ns := framework.CreateNamespaceOrDie(testClient, namespaceName, t)
+	defer framework.DeleteNamespaceOrDie(testClient, ns, t)
+
+	// create a node
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+			Annotations: map[string]string{
+				util.ControllerManagedAttachAnnotation: "true",
+			},
+		},
+	}
+	node, err := testClient.CoreV1().Nodes().Create(context.TODO(), node, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to created node : %v", err)
+	}
+
+	// create fake pods with volumes
+	pod := fakePodWithVol(namespaceName)
+	if _, err := testClient.CoreV1().Pods(ns.Name).Create(context.TODO(), pod, metav1.CreateOptions{}); err != nil {
+		t.Errorf("Failed to create pod : %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go informers.Core().V1().Nodes().Informer().Run(ctx.Done())
+	podInformer := informers.Core().V1().Pods().Informer()
+	// start the informer
+	informers.Start(ctx.Done())
+	informers.WaitForCacheSync(ctx.Done())
+
+	// run the controllers
+	go ctrl.Run(ctx)
+	go pvCtrl.Run(ctx)
+	go gcCtrl.Run(ctx)
+
+	waitToObservePods(t, podInformer, 1)
+
+	// wait for volume to be attached
+	for i := 0; i < 10; i++ {
+		node, err = testClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("Failed to get the node : %v", err)
+		}
+		if len(node.Status.VolumesAttached) > 1 {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+	if len(node.Status.VolumesAttached) < 1 {
+		t.Logf("failed to attach volume for pod %s on node %s", pod.Name, node.Name)
+	}
+	// Patch the node to mark the volume in use as attach-detach controller verifies if safe to detach the volume
+	// based on that.
+	node.Status.VolumesInUse = append(node.Status.VolumesInUse, "kubernetes.io/mock-provisioner/fake-mount")
+	node, err = testClient.CoreV1().Nodes().UpdateStatus(context.TODO(), node, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("error in updating node status: %v", err)
+	}
+	// We need to mimic a non-graceful node shutdown by marking the node to be not ready but the created node
+	// does not have any condition so it will be `Not Ready` and hence a patch of Node to make it `Not Ready`
+	// is not required.
+
+	// Delete the pod with grace period time so that it is stuck in terminating state
+	gracePeriod := int64(300)
+	err = testClient.CoreV1().Pods(namespaceName).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriod,
+	})
+	if err != nil {
+		t.Fatalf("error in deleting pod: %v", err)
+	}
```

Comment:
Please add comment like “verify pod.ObjectMeta.DeletionTimestamp is not nil. This means pod is in ‘Terminating’ state”.

Reply:
done